### ISSUE TITLE
Need instructions for configuring project to use Debugs IDs

### ISSUE THREAD

**Issue by markholland** [getsentry/sentry-react-native#3388]
On Sentry.io it says that Debug IDs is the recommended way to use Sentry.

![Screenshot 2023-11-13 at 11 35 25](https://github.com/getsentry/sentry-react-native/assets/1764217/6518f573-93fe-4cf4-9d50-35d7b2ee937c)

It appears to have been introduced in [5.11.0](https://github.com/getsentry/sentry-react-native/releases/tag/5.11.0)

But the instructions there are pretty vague and seem incomplete. Also the links provided don't mention Debug IDs in anyway.

The first step has you import a package and create an object inside `metro.config.js` but then no information on what comes next.

![Screenshot 2023-11-13 at 14 49 17](https://github.com/getsentry/sentry-react-native/assets/1764217/f4e57d05-d508-4860-83ca-ddce01a1bc85)

The next step has you update Xcode project scripts but includes paths to non-existent `scripts/` directory. Maybe it needs to be `../node_modules/@sentry/react-native/scripts/sentry-xcode.sh` and `../node_modules/@sentry/react-native/scripts/sentry-xcode-debug-files.sh`?

![Screenshot 2023-11-13 at 14 49 25](https://github.com/getsentry/sentry-react-native/assets/1764217/1cadbb21-4ee1-45be-a23a-4f0c50796c2d)

**Comment by krystofwoldrich** [comment]
Hi @markholland,
thank you for the message,
yes the paths should be `../node_modules/@sentry/react-native/scripts/sentry-xcode.sh` and `../node_modules/@sentry/react-native/scripts/sentry-xcode-debug-files.sh`.

These are the only two needed steps. The Serializer creates the DebugID and the new script ensures it's uploaded.

**Comment by markholland** [comment]
Hi @krystofwoldrich and thanks for the reply.

Thanks for the clarification but I think there may still be some missing configuration steps. Once setting the correct paths it does indeed build and send errors to Sentry but once on Sentry.io I'm having some problems with the result.

I've provided link to the issues and hopefully you can help with what might be required to resolve each problem.

On iOS it seems to have failed to apply any source mapping despite everything looking correct in the project settings for this build https://catenda-as.sentry.io/issues/4629282026/events/4cb51c0aad144900955d78e713939ed4/

**Source Maps for this build**
![Screenshot 2023-11-14 at 09 48 38](https://github.com/getsentry/sentry-react-native/assets/1764217/65e605f3-d576-4216-adae-3e2faa1bcae4)

On Android, https://catenda-as.sentry.io/issues/4309830290/events/b5c1e86dd72c4ccfab272cc30f45bd47/ , source maps are correctly applied but it's still saying Debug IDs isn't correctly configured and the link provided doesn't relate to React Native. Maybe these new scripts don't include the `inject` flag as described [here](https://docs.sentry.io/platforms/javascript/sourcemaps/uploading/cli/#3-inject-debug-ids-into-artifacts)?

![Screenshot 2023-11-14 at 09 46 15](https://github.com/getsentry/sentry-react-native/assets/1764217/fa475060-83f5-44ab-8356-b48b5cf8b8fb)

**Comment by krystofwoldrich** [comment]
@markholland I can confirm the events are coming without DebugIDs.

Could you share the output of the `Bundle React Native Code and Images` build phase for iOS and the `createBundleReleaseJsAndAssets_SentryUpload_*` Gradle Task?

**Comment by markholland** [comment]
Here's the Android logs. I'll have to do another build for iOS as fast lane hides the output by default

```
> Task :app:createBundleReleaseJsAndAssets_SentryUpload_no.catenda.bimsyncmobile@3.3.0+511_android-511
[14:16:24]: ▸ Copy `debugId` from packager source map to Hermes source map...
[14:16:24]: ▸ Packager source map does not have `debugId`.
[14:16:24]: ▸ Check generated source map for Debug ID: /bitrise/src/android/app/build/generated/sourcemaps/react/release/index.android.bundle.map does not contain 'debugId' nor 'debug_id'.
[14:16:24]: ▸ Sentry Source Maps upload will include the release name and dist.
[14:16:24]: ▸ Sentry-CLI arguments: [/bitrise/src/node_modules/@sentry/cli/bin/sentry-cli, react-native, gradle, --bundle, /bitrise/src/android/app/build/ASSETS/createBundleReleaseJsAndAssets/index.android.bundle, --sourcemap, /bitrise/src/android/app/build/generated/sourcemaps/react/release/index.android.bundle.map, --release, no.catenda.bimsyncmobile@3.3.0+511, --dist, android-511]
[14:16:24]: ▸ INFO    2023-11-13 14:16:22.813420358 +00:00 Loaded file referenced by SENTRY_PROPERTIES (/bitrise/src/android/sentry.properties)
[14:16:24]: ▸ Processing react-native sourcemaps for Sentry upload.
[14:16:24]: ▸ > Analyzing 2 sources
[14:16:24]: ▸ > Rewriting sources
[14:16:24]: ▸ > Adding source map references
[14:16:24]: ▸ Uploading sourcemaps for release no.catenda.bimsyncmobile@3.3.0+511 distribution android-511
[14:16:24]: ▸ > Bundled 2 files for upload
[14:16:24]: ▸ > Bundle ID: 5956cef2-6722-5efe-8eb9-0d5230f92af2
[14:16:25]: ▸ > Uploaded files to Sentry
[14:16:26]: ▸ > File upload complete (processing pending on server)
[14:16:26]: ▸ > Organization: catenda-as
[14:16:26]: ▸ > Project: bimsync-mobile
[14:16:26]: ▸ > Release: no.catenda.bimsyncmobile@3.3.0+511
[14:16:26]: ▸ > Dist: android-511
[14:16:26]: ▸ > Upload type: artifact bundle
[14:16:26]: ▸ Source Map Upload Report
[14:16:26]: ▸ Minified Scripts
[14:16:26]: ▸ ~/index.android.bundle (sourcemap at index.android.bundle.map)
[14:16:26]: ▸ Source Maps
[14:16:26]: ▸ ~/index.android.bundle.map
```

Considering the output above already states that it can't find debug ids I'll also share our `metro.config.js` as I expect that to be the next step.

```
const {
  createSentryMetroSerializer,
} = require('@sentry/react-native/dist/js/tools/sentryMetroSerializer');
// eslint-disable-next-line no-unused-vars
const config = { serializer: createSentryMetroSerializer() };

// maxWorker & transformer: Fix a sentry error on generating sourcemaps https://github.com/getsentry/sentry-react-native/issues/484#issuecomment-534646955
// resolver: Add mjs for make-plural https://github.com/eemeli/make-plural/issues/15#issuecomment-672854945
module.exports = {
  maxWorkers: 4,
  resolver: {
    sourceExts: ['js', 'jsx', 'json', 'mjs', 'ts', 'tsx'],
  },
  transformer: {
    getTransformOptions: async () => ({
      transform: {
        experimentalImportSupport: false,
        inlineRequires: false,
      },
    }),
  },
};
```

**Comment by markholland** [comment]
The source maps not working on iOS was due to something in Sentry scripts not supporting projects/targets with spaces. After removing the space source maps appear correctly but get the same Debug IDs configuration issue in Sentry.io as shared previously.

I've also noticed that it's complaining under the Releases tab even though the release does match

![Screenshot 2023-11-14 at 12 27 06](https://github.com/getsentry/sentry-react-native/assets/1764217/5bad2c36-0dca-433f-a25e-d4c4a55c52fc)

![Screenshot 2023-11-14 at 12 27 17](https://github.com/getsentry/sentry-react-native/assets/1764217/23fabcd9-a0ca-4992-8367-cce05c372bc5)

Here's the output of `Bundle React Native Code and Images`, I tried to remove the unrelated parts otherwise it's 2000k lines

```
PhaseScriptExecution Bundle\ React\ Native\ code\ and\ images /Users/markholland/Library/Developer/Xcode/DerivedData/catenda-site-gjylplrpdgescehkebdaxrkfihqv/Build/Intermediates.noindex/catenda-site.build/Debug-iphonesimulator/catenda-site.build/Script-00DD1BFF1BD5951E006B06BC.sh (in target 'catenda-site' from project 'catenda-site')
    cd /Users/markholland/Documents/catenda/bimsync-mobile/ios
    /bin/sh -c /Users/markholland/Library/Developer/Xcode/DerivedData/catenda-site-gjylplrpdgescehkebdaxrkfihqv/Build/Intermediates.noindex/catenda-site.build/Debug-iphonesimulator/catenda-site.build/Script-00DD1BFF1BD5951E006B06BC.sh

+ WITH_ENVIRONMENT=../node_modules/react-native/scripts/xcode/with-environment.sh
+ REACT_NATIVE_XCODE=../node_modules/react-native/scripts/react-native-xcode.sh
+ /bin/sh -c '../node_modules/react-native/scripts/xcode/with-environment.sh "/bin/sh ../node_modules/@sentry/react-native/scripts/sentry-xcode.sh ../node_modules/react-native/scripts/react-native-xcode.sh"'
Node found at: /usr/local/bin/node
+ '[' -z '' ']'
+ export SENTRY_PROPERTIES=sentry.properties
+ SENTRY_PROPERTIES=sentry.properties
+ '[' -z '' ']'
+ export SOURCEMAP_FILE=/Users/markholland/Library/Developer/Xcode/DerivedData/catenda-site-gjylplrpdgescehkebdaxrkfihqv/Build/Intermediates.noindex/catenda-site.build/Debug-iphonesimulator/catenda-site.build/DerivedSources/main.jsbundle.map
+ SOURCEMAP_FILE=/Users/markholland/Library/Developer/Xcode/DerivedData/catenda-site-gjylplrpdgescehkebdaxrkfihqv/Build/Intermediates.noindex/catenda-site.build/Debug-iphonesimulator/catenda-site.build/DerivedSources/main.jsbundle.map
+ '[' -z '' ']'
+ SENTRY_CLI_EXECUTABLE=../node_modules/@sentry/cli/bin/sentry-cli
+ REACT_NATIVE_XCODE=../node_modules/react-native/scripts/react-native-xcode.sh
+ [[ '' != true ]]
+ [[ -z '' ]]
+ NO_AUTO_RELEASE=--no-auto-release
+ ARGS='--no-auto-release  '
+ REACT_NATIVE_XCODE_WITH_SENTRY='"../node_modules/@sentry/cli/bin/sentry-cli" react-native xcode --no-auto-release   "../node_modules/react-native/scripts/react-native-xcode.sh"'
+ '[' '' '!=' true ']'
+ /bin/sh -c '"../node_modules/@sentry/cli/bin/sentry-cli" react-native xcode --no-auto-release   "../node_modules/react-native/scripts/react-native-xcode.sh"'
  INFO    2023-11-14 12:32:07.245336 +01:00 Loaded file referenced by SENTRY_PROPERTIES (sentry.properties)
+ DEST=/Users/markholland/Library/Developer/Xcode/DerivedData/catenda-site-gjylplrpdgescehkebdaxrkfihqv/Build/Products/Debug-iphonesimulator/catenda-site.app
+ [[ ! -n '' ]]
+ [[ Debug = *Debug* ]]
+ [[ ! iphonesimulator == *simulator ]]
+ [[ -n '' ]]
+ case "$CONFIGURATION" in
+ [[ iphonesimulator == *simulator ]]
+ [[ -n '' ]]
+ echo 'Skipping bundling in Debug for the Simulator (since the packager bundles for you). Use the FORCE_BUNDLING flag to change this behavior.'
Skipping bundling in Debug for the Simulator (since the packager bundles for you). Use the FORCE_BUNDLING flag to change this behavior.
+ exit 0
+ '[' -z '' ']'
+ SENTRY_COLLECT_MODULES=../../scripts/collect-modules.sh
+ '[' -f ../../scripts/collect-modules.sh ']'
```

**Comment by markholland** [comment]
I adjusted `metro.config.js` to the following as maybe that was the intention and just was'nt detailed in the instructions but the outcome is the same.

```
const {
  createSentryMetroSerializer,
} = require('@sentry/react-native/dist/js/tools/sentryMetroSerializer');

// maxWorker & transformer: Fix a sentry error on generating sourcemaps https://github.com/getsentry/sentry-react-native/issues/484#issuecomment-534646955
module.exports = {
  maxWorkers: 4,
  resolver: {
    sourceExts: ['js', 'jsx', 'json', 'mjs', 'ts', 'tsx'],
  },
  serializer: createSentryMetroSerializer(),
  transformer: {
    getTransformOptions: async () => ({
      transform: {
        experimentalImportSupport: false,
        inlineRequires: false,
      },
    }),
  },
};
```

**Comment by krystofwoldrich** [comment]
@markholland Thank you for all the logs and the forward-thinking, that really helps.

I see what happened. There was one property missing (`customSerializer`) in the changelog example.

```
  serializer: {
    customSerializer: createSentryMetroSerializer(),
  },
```

Your config with the serializer:
```
const {
  createSentryMetroSerializer,
} = require('@sentry/react-native/dist/js/tools/sentryMetroSerializer');

// maxWorker & transformer: Fix a sentry error on generating sourcemaps https://github.com/getsentry/sentry-react-native/issues/484#issuecomment-534646955
module.exports = {
  maxWorkers: 4,
  resolver: {
    sourceExts: ['js', 'jsx', 'json', 'mjs', 'ts', 'tsx'],
  },
  serializer: {
    customSerializer: createSentryMetroSerializer(),
  },
  transformer: {
    getTransformOptions: async () => ({
      transform: {
        experimentalImportSupport: false,
        inlineRequires: false,
      },
    }),
  },
};
```

**Comment by markholland** [comment]
@krystofwoldrich thanks for the reply.

Looks like that was what was missing as everything appears to be correct in Sentry.io now.

Hopefully y'all be able to add more detailed instructions to the documentation website so that this is easier to get working for others.

One final question, now that we're not relying on releases thanks to Debug IDs do we still need to set the environment variables SENTRY_DIST & SENTRY_RELEASE during the builds? Just trying to clean up anything that is no longer required.

Thanks for the help.

**Comment by pranshuchittora** [comment]
Thank you so much for addressing this issue. I spent an entire day trying to understand why the sourcemaps were breaking on Sentry. Initially, I followed the changes outlined in the release changelog (the old ones), only to realize that this was the root cause of the problem.

I noticed that the changelog has been edited and updated with the correct configuration. It would be helpful to highlight this correction, perhaps in the latest release notes or another prominent location. This way, individuals encountering the same error or using an incorrect metro config can easily rectify the issue.
